src/arithmetic.rs

```diff
@@ -430,3 +430,44 @@ fn log2_floor(num: usize) -> u32 {
 
     pow
 }
+
+/// Returns coefficients of an n - 1 degree polynomial given a set of n points and their evaluations
+pub fn lagrange_interpolate<F: Field>(points: &[F], evals: &[F]) -> Vec<F> {
+    assert_eq!(points.len(), evals.len());
+    if points.len() == 1 {
+        // Constant polynomial
+        return vec![evals[0]];
+    } else {
+        let mut interpolation_polys = vec![];
+        for (j, x_j) in points.iter().enumerate() {
+            let mut tmp: Vec<F> = Vec::with_capacity(points.len() + 1);
+            let mut product = Vec::with_capacity(points.len() + 1);
+            tmp.push(F::one());
+            for (k, x_k) in points.iter().enumerate() {
+                if k != j {
+                    // Compute (x_j - x_k)^(-1)
+                    let denom = (*x_j - x_k).invert().unwrap();
+                    product.resize(tmp.len() + 1, F::zero());
+                    for (i, (a, b)) in tmp
+                        .iter()
+                        .chain(std::iter::once(&F::zero()))
+                        .zip(std::iter::once(&F::zero()).chain(tmp.iter()))
+                        .enumerate()
+                    {
+                        product[i] = *a * (-denom * x_k) + *b * denom;
+                    }
+                    std::mem::swap(&mut tmp, &mut product);
+                }
+            }
```

Comment:
Add an assertion here that the capacity we initialized `product`/`tmp` to originally matches the actual length at the end of the loop.